## Selfish Genes

Quanta puzzle 2017-09-14

https://www.quantamagazine.org/are-genes-selfish-or-cooperative-20170914/

In [1]:
# Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Problem 1

In [32]:
# Set the initial distribution of genotypes
df = pd.DataFrame([[0.3, 0.6, 0.1]],
                  columns=["AA", "Aa", "aa"])

# df.loc[1] = [1,23,4]

def next_gen(curr_gen):
    """
    Returns the distribution of genotypes for a given initial distribution
    Accepts a 3x1 list, containing the generation number, and probabilities for the three genotypes
    Returns a 3x1 list, containing the same for the next generation
    """
    
    # Unpack the list, for legibility
    pAA, pAa, paa = curr_gen
    
    next_gen = [0,0,0] # This line can probably be improved
    
    # Compute proportion of AAs in next generation (next_gen[0])
    next_gen[0] =  pAA * pAA # AA meets AA
    next_gen[0] += pAA * pAa # AA meets Aa
    next_gen[0] += pAa * pAa * 0.25 # Aa meets Aa
    
    # Compute proportion of Aas in next generation (next_gen[1])
    next_gen[1] =  pAA * pAa # AA meets Aa
    next_gen[1] += pAA * paa * 2 # AA meets aa
    next_gen[1] += pAa * pAa * 0.5 # Aa meets Aa 
    next_gen[1] += pAa * paa # Aa meets aa 
    
    # Compute proportion of aas in next generation (next_gen[2])
    next_gen[2] =  pAa * pAa * 0.25 # Aa meets Aa
    next_gen[2] += pAa * paa # Aa meets aa
    next_gen[2] += paa * paa # aa meets aa

    map(lambda x: round(x, 10), next_gen)
    
    return next_gen

# Simulate 10 generations
for x in range(10):
    df.loc[x + 1] = next_gen(df.iloc[x])

# Add a sum of the proportions, to ensure it is 100%
df['Sum'] = df['AA'] + df['Aa'] + df['aa']

# Compute the proportion of the population by allele
df['Amount of A'] = df['AA'] * 2 + df['Aa']
df['Amount of a'] = df['aa'] * 2 + df['Aa']

# Print out the entire DataFrame
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df)
    
df.to_csv("set1", sep="\t")

      AA    Aa    aa  Sum  Amount of A  Amount of a
0   0.30  0.60  0.10  1.0          1.2          0.8
1   0.36  0.48  0.16  1.0          1.2          0.8
2   0.36  0.48  0.16  1.0          1.2          0.8
3   0.36  0.48  0.16  1.0          1.2          0.8
4   0.36  0.48  0.16  1.0          1.2          0.8
5   0.36  0.48  0.16  1.0          1.2          0.8
6   0.36  0.48  0.16  1.0          1.2          0.8
7   0.36  0.48  0.16  1.0          1.2          0.8
8   0.36  0.48  0.16  1.0          1.2          0.8
9   0.36  0.48  0.16  1.0          1.2          0.8
10  0.36  0.48  0.16  1.0          1.2          0.8


The distribution of genotypes stabilises after the first generation - it remains the same for all subsequent generations. (This can be tested for other initial distributions. Using the code above, there will be changes after about 30 generations, but this seems to be due to the accumulation of rounding errors.)

### Problem 2

Looking at the three initial distributions suggested, we see:



| Generation | AA   | Aa   | aa   | Sum | Amount of A | Amount of a |
|------------|------|------|------|-----|-------------|-------------|
| 0          | 0.3  | 0.6  | 0.1  | 1.0 | 1.2         | 0.8         |
| 1          | 0.36 | 0.48 | 0.16 | 1.0 | 1.2         | 0.8         |


| Generation | AA   | Aa   | aa   | Sum | Amount of A | Amount of a |
|------------|------|------|------|-----|-------------|-------------|
| 0          | 0.2  | 0.8  | 0.0  | 1.0 | 1.2         | 0.8         |
| 1          | 0.36 | 0.48 | 0.16 | 1.0 | 1.2         | 0.8         |


| Generation | AA   | Aa   | aa   | Sum | Amount of A | Amount of a |
|------------|------|------|------|-----|-------------|-------------|
| 0          | 0.7  | 0.2  | 0.1  | 1.0 | 1.6         | 0.4         |
| 1          | 0.64 | 0.32 | 0.04 | 1.0 | 1.6         | 0.4         |

In all cases, the distribution after generation 1 remains the same as that in generation 1.
More interestingly, the distribution of alleles does not change at all; whatever was present in the initial generation is retained.
